### PR TITLE
Switch To Calculating Networking Information Directly For Performance

### LINKED ISSUE #132: Valid IPv6 input may cause 'ipsubnet' filter to attempt to create 2 ^ 128 values
##### SUMMARY

The `"2600:1f1c:1b3:8f00::/56" | ipsubnet(120, 0)` filter never returns. Internally, the filter invokes the `netaddr.subnet` function, which attempts to create 18446744073709551616 values.

The `ipsubnet` filter creates a IPNetwork with "2600:1f1c:1b3:8f00::/56" and prefix length 120.
The `ipsubnet` filter invokes the `netaddr.subnet` function at https://github.com/ansible-collections/ansible.utils/blob/main/plugins/filter/ipsubnet.py#L283
)

```python
return str(list(value.subnet(query))[index])
```

In the `netaddr` module, the [subnet()](https://github.com/netaddr/netaddr/blob/master/netaddr/ip/__init__.py#L1310-L1316) function calculates the max number of subnets:

```python
width = self._module.width
max_subnets = 2 ** (width - self.prefixlen) // 2 ** (width - prefixlen)
```

If the network prefix is /56 and the prefix length is /120, then max_subnets = 18446744073709551616.

In the worst case, the network address is /0 and the prefix length is 128, then max_subnets = 2 ^ 128 // 2 ^ 0, which is 2 ^ 128. If the `count` argument is set to 0, this will cause unbounded CPU usage and memory allocation.
So `netaddr.subnet` never returns.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ipsubnet filter

##### ANSIBLE VERSION

I have pasted the output below, but the problem is reproduced when running unit tests in the `main` branch of `ansible.utils` module as of 01/24/2022.

```paste below
ansible [core 2.12.1]
  config file = None
  configured module search path = ['/home/serosset/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/serosset/.local/lib/python3.8/site-packages/ansible
  ansible collection location = /home/serosset/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/rh/rh-python38/root/usr/local/bin/ansible
  python version = 3.8.11 (default, Sep  1 2021, 12:33:46) [GCC 9.3.1 20200408 (Red Hat 9.3.1-2)]
  jinja version = 3.0.3
  libyaml = True
```

##### COLLECTION VERSION

```paste below

```

Because the problem is exposed when running the test_ipsubnet.py unit tests inside a docker image, the applicable collections and configuration is what's inside the docker container. I'm not sure how to get that information. But see summary, I think the problem can be root caused to a specific line in the `ipsubnet` function in the main branch.
```
ansible-test units --python=3.10 --docker -vvvv tests/unit/plugins/filter/test_ipsubnet.py
```

##### CONFIGURATION

```paste below

```

##### STEPS TO REPRODUCE

Checkout the code from #131, then run the following command:
```
# ansible-test units --python=3.10 --docker -vvvv tests/unit/plugins/filter/test_ipsubnet.py
```

The following filter never completes: `"2600:1f1c:1b3:8f00::/56" | ipsubnet(120, 0)`

The problem gets worse when the network prefix has a smaller value and the first argument of `ipsubnet` has a higher value. In the worst case, this would cause 2 ^ 128 iterations.

```yaml

```

##### EXPECTED RESULTS

`"2600:1f1c:1b3:8f00::/56" | ipsubnet(120, 0)` should return `2600:1f1c:1b3:8f00::/120`
`"2600:1f1c:1b3:8f00::/56" | ipsubnet(120, 4)` should return `2600:1f1c:1b3:8f00::400/120`

This could be calculated in constant time. Instead, the `ipsubnet` filter attempts to create all possible subnets, then get the subnet at index X. However, the list of all subnets could be 2 ^ 128 in the worst case.

##### ACTUAL RESULTS

The `ipsubnet` filter never returns. It iterates through a very large number of elements and never returns.

```paste below

```

### PR DESCRIPTION
**Pull request by timway**
##### SUMMARY

The `netaddr` library provides a convenience method that provides a generator for subnets of an `IPNetwork` object. This was used to solve several of the use cases of this filter. Unfortunately this only remains usable with small networks or limited gaps between the original network and the size of the requested subnets. As we move into IPv6 this will become increasingly problematic (/64 is normal and it may want to be split into very small subnets > 120).

I added several new tests to cover some edge cases I uncovered while writing it. I think I have all of the `black` related changes in as well although the `tests/unit/plugins/filter/test_ipsubnet.py` was requiring more changes in my local `black` so I'm going to have `zuul` run through it and see if it pops there too before I add a commit for those formatting related changes.

Fixes #132

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ipsubnet

##### ADDITIONAL INFORMATION

Try this task before and after these commits. In the current version it will hang, this code will return extremely quickly.
```paste below
    - ansible.builtin.debug:
        msg: "{{ '7fff:ffff:ffff:ffff:ffff:ffff:ffff:fffe/127' | ansible.utils.ipsubnet('9000::/1') }}"
```

### CODE REVIEW

**Comment by timway** [comment]
@ashwini-mhatre added tests to main around when I was writing this PR up and I didn't see them land into main. I'll finish evaluating the new tests and get everything passing again as time allows. I am able to replicate the failures locally now.

**Comment by timway** [comment]
@ashwini-mhatre this is ready to be reviewed. I noticed you have some changes in #149 that touch some of the same files so that may cause some conflicts depending on the order this gets merged. I also made a change to the test you added in `test_ipaddr.py` to better account for the difference between the Python versions when the number of subnets inside of a network is requested and the query is for a supernet of the subnet. It now will reliably raise an error in addition to the performance work that spurred the request.

I've also removed a lot of the `try..catch` logic that was used to return false in several places with clear tests (if) of the conditions that it should return false under.

File: changelogs/fragments/ipsubnet_performance.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+bugfixes:
```
Comment (ashwini-mhatre) [inline changelogs/fragments/ipsubnet_performance.yaml:2]: This will be minor_change as you are updating type of arg from str to int

**Review by ashwini-mhatre — COMMENTED** [review]
can you please add some integration tests?

**Comment by ashwini-mhatre** [comment]
@timway could you please add some integration tests?

**Comment by ashwini-mhatre** [comment]
@cidrblock if possible could you please review this PR?

File: changelogs/fragments/ipsubnet_performance.yaml (hunk 1 above)
Comment (timway) [reply changelogs/fragments/ipsubnet_performance.yaml:2]: @ashwini-mhatre I've split out the changes so the performance work is under bugfix and the parameter change is under `minor_changes`.

**Comment by timway** [comment]
> @timway could you please add some integration tests?

@ashwini-mhatre, I've added a pair of tests that covers the original performance issue.

**Review by ashwini-mhatre — APPROVED** [review]

**Comment by goneri** [comment]
recheck

**Comment by goneri** [comment]
recheck

**Comment by ashwini-mhatre** [comment]
recheck

File: plugins/filter/ipsubnet.py [hunk 2]
```diff
@@ -252,73 +252,88 @@ def _ipsubnet(*args, **kwargs):
...
-    query_string = str(query)
+    elif vtype == "address":
+        v = ipaddr(value, "cidr")
+    elif vtype == "network":
+        v = ipaddr(value, "subnet")
+    value = netaddr.IPNetwork(v).cidr
+    vtype = ipaddr(value, "type")
+
     if not query:
         return str(value)
```
Comment (Qalthos) [inline plugins/filter/ipsubnet.py:269]: I realize this isn't even something you've touched, but while we're in here we should always be returning `to_text()` instead of `str()`

**Review by Qalthos — CHANGES_REQUESTED** [review]
> I think we'd get or want a different function in that api as "subnet" very clearly is meant to be a list of the subnets. something like 'subnet_at(\<size\>, \<position\>)'

If you can reorganize this to create the `subnet_at` function in the ipsubnet plugin file, that should clean up the ipsubnet filter's code and be able to be moved if we want it somewhere else (or be replaced if something similar enough lands in netaddr)

**Review by Qalthos — APPROVED** [review]
